src/pages/_app.tsx

```diff
@@ -1,5 +1,12 @@
-import type { AppProps } from "next/app";
+import ObserverContextProvider from '@/context/ObserverContext';
+import type { AppProps } from 'next/app';
+
+import '../styles/index.css';
 
 export default function App({ Component, pageProps }: AppProps) {
-  return <Component {...pageProps} />;
+	return (
+		<ObserverContextProvider>
```

Comment:
> **issue:** It will cause whole react tree re-rendered each time it state changes.

No it's won't. 
Context provider don't causing whole react tree re-render. 
It just provide the global state to all components. 
If specific component use this state, then this specific component will re-render, but not whole tree.